contracts/coordinator/src/state.rs

```diff
@@ -12,17 +12,79 @@ pub struct Config {
 
 pub const CONFIG: Item<Config> = Item::new("config");
 
-type ProverAddress = Addr;
+pub type ProverAddress = Addr;
 
 pub const PROVER_PER_CHAIN: Map<ChainName, ProverAddress> = Map::new("prover_per_chain");
 
-// TODO: migrate this?
-pub const ACTIVE_VERIFIER_SET_FOR_PROVER: Map<ProverAddress, VerifierSet> =
-    Map::new("active_prover_verifier_set");
-
-type ChainNames = HashSet<ChainName>;
-type VerifierAddress = Addr;
+pub type ChainNames = HashSet<ChainName>;
+pub type VerifierAddress = Addr;
 pub const CHAINS_OF_VERIFIER: Map<VerifierAddress, ChainNames> = Map::new("chains_of_verifier");
 
-pub const NEXT_VERIFIER_SET_FOR_PROVER: Map<ProverAddress, VerifierSet> =
-    Map::new("next_prover_verifier");
+pub struct VerifierSetIndex<'a> {
+    pub by_verifier: MultiIndex<'a, Addr, VerifierProverRecord, (Addr, Addr)>,
+}
+
+impl<'a> IndexList<VerifierProverRecord> for VerifierSetIndex<'a> {
+    fn get_indexes(&self) -> Box<dyn Iterator<Item = &dyn Index<VerifierProverRecord>> + '_> {
+        let v: Vec<&dyn Index<VerifierProverRecord>> = vec![&self.by_verifier];
+        Box::new(v.into_iter())
+    }
+}
+
+#[cw_serde]
+pub struct VerifierProverRecord {
+    pub prover: ProverAddress,
+    pub verifier: VerifierAddress,
+}
+
+pub const VERIFIER_PROVER_INDEXED_MAP: IndexedMap<
+    (Addr, Addr),
+    VerifierProverRecord,
+    VerifierSetIndex,
+> = IndexedMap::new(
+    "verifier_prover_map",
+    VerifierSetIndex {
+        by_verifier: MultiIndex::new(
+            |_pk: &[u8], d| d.verifier.clone(),
+            "verifier_prover_map",
+            "verifier_prover_map_by_verifier",
+        ),
+    },
+);
+
+pub fn update_verifier_set_for_prover(
+    storage: &mut dyn Storage,
+    prover_address: ProverAddress,
+    new_verifiers: HashSet<VerifierAddress>,
+) -> Result<(), ContractError> {
+    let existing_verifiers = VERIFIER_PROVER_INDEXED_MAP
+        .prefix(prover_address.clone())
+        .keys(storage, None, None, Order::Ascending)
+        .filter_map(Result::ok)
+        .collect::<HashSet<VerifierAddress>>();
+
+    for verifier in existing_verifiers.difference(&new_verifiers) {
+        if VERIFIER_PROVER_INDEXED_MAP
+            .remove(storage, (prover_address.clone(), verifier.clone()))
+            .is_err()
+        {
+            return Err(ContractError::NoSuchVerifierRegisteredForProver(
+                verifier.clone(),
+                prover_address,
+            ));
+        }
```

Comment:
that looks like it should be possible to solve with `change_context`/`map_err` and `?`